tests/unit/charms/tls_certificates_interface/v2/test_tls_certificates_v2_provides.py

```diff
@@ -185,6 +185,92 @@ def test_given_certificate_for_csr_already_in_relation_data_when_on_relation_cha
 
         patch_certificate_creation_request.assert_not_called()
 
+    @patch(
+        f"{LIB_DIR}.CertificatesProviderCharmEvents.certificate_creation_request",
+        new_callable=PropertyMock,
+    )
+    def test_given_certificate_for_csr_already_in_relation_data_when_on_relation_changed_and_unit_not_leader_then_certificate_creation_request_is_emitted(  # noqa: E501
+        self, patch_certificate_creation_request
+    ):
+        relation_id = self.create_certificates_relation_with_1_remote_unit()
+        self.harness.set_leader(is_leader=True)
+        csr = "whatever csr"
+        provider_app_data = {
+            "certificates": json.dumps(
+                [
+                    {
+                        "certificate_signing_request": csr,
+                        "certificate": "whatever cert",
+                        "ca": "whatever ca",
+                        "chain": ["whatever cert 1", "whatever cert 2"],
+                    }
+                ]
+            )
+        }
+        self.harness.update_relation_data(
+            relation_id=relation_id,
+            app_or_unit=self.harness.charm.app.name,
+            key_values=provider_app_data,
+        )
+
+        requirer_unit_data = {
+            "certificate_signing_requests": json.dumps(
+                [
+                    {
+                        "certificate_signing_request": csr,
+                    }
+                ]
+            )
+        }
+        self.harness.set_leader(is_leader=False)
```

Comment:
Why are we calling `set_leader` here?

Reply:
We are testing unit is not leader when relation changed. This is why we are setting leader False.

Reply:
But it was leader at the beginning of the test, no?

Reply:
If we check the test definition CSR is already in the relation data. To require a leader unit to  prepare the conditions that this test needs. Then we are testing a relation changed situation with a non leader unit.

Reply:
Yeah, I get it now, but in that case we are testing that we request the creation of certificate for a csr that already has a certificate, no?
Is that an expected behaviour? This is a change compared to what the library was doing before.
I think the root of this is that in the `on_relation_changed` in the provider we require the unit to be leader to get `provider_relation_data` but we don't have that requirement to get the `requirer_relation_data`
I think in this case if a unit is not leader it should not do anything.
If we use the `self._load_relation_data` in both cases we will get to that, but I again find it a bit hard to follow, I'd suggest using guards in place where needed and keeping one `_load_relation_data` or using `self._load_relation_data` only in the provider.

Reply:
Yes, we are testing that we request the creation of certificate for a crs that already has a certificate. If the relation changed and provider unit is not leader, if can not get the certificate which is already there as we put a guard for non-leader unit. You can see the there is no guard in the requirer  part and we are using self._load_relation_data  only in the provider side if  we try to app data bag. As we can not that certificate because of guard, code process it like it does not exist.  This causes to trigger another certificate creation request.  You think this part is problematic, right ? A non leader unit should not cause to any action.

Reply:
I think the problematic part is in the provider in the `_on_relation_changed`. 
The logic there is to filter CSRs that already got their certificates, now we don't do that if the provider unit is not leader, we just issue a new certificate for them.
That part needs a revisit. In my opinion it should be like it was before, no certificate should be issued to a CSR with a certificate (can be done by using guards in provider `_on_relation_changed`).
Let me know what you think and if you have a different suggestion.

Reply:
This is very good catch. I fixed it in the latest patch. The code was comparing certificates_signing_requests in provider_csrs and requirer_csrs. However provider_csrs  was not reflecting the truth so this situation happened. I fixed this problem adding another check on_relation_changed_method before processing to those step. Now a non-leader unit does nothing.